cpp/src/KIM_FilesystemPath.hpp

```diff
@@ -0,0 +1,203 @@
+//
+// CDDL HEADER START
+//
+// The contents of this file are subject to the terms of the Common Development
+// and Distribution License Version 1.0 (the "License").
+//
+// You can obtain a copy of the license at
+// http://www.opensource.org/licenses/CDDL-1.0.  See the License for the
+// specific language governing permissions and limitations under the License.
+//
+// When distributing Covered Code, include this CDDL HEADER in each file and
+// include the License file in a prominent location with the name LICENSE.CDDL.
+// If applicable, add the following below this CDDL HEADER, with the fields
+// enclosed by brackets "[]" replaced with your own identifying information:
+//
+// Portions Copyright (c) [yyyy] [name of copyright owner]. All rights reserved.
+//
+// CDDL HEADER END
+//
+
+//
+// Copyright (c) 2016--2019, Regents of the University of Minnesota.
+// All rights reserved.
+//
+// Contributors:
+//    Ryan S. Elliott
+//    Alexander Stukowski
+//
+
+//
+// Release: This file is part of the kim-api.git repository.
+//
+
+
+#ifndef KIM_FILESYSTEM_PATH_HPP_
+#define KIM_FILESYSTEM_PATH_HPP_
+
+#include <cstdlib>
+#include <iostream>
+#include <vector>
+
+// If available on the local platform, use the std::filesystem::path class
+// for cross-platform file path handling introduced with C++17. Otherwise,
+// fall back to a minimal implementation based on std::string.
+#if __cplusplus >= 201703L && defined(__has_include) && __has_include(<filesystem>)
+#define KIM_API_USE_FILESYSTEM_LIBRARY 1
+#include <filesystem>
+#else
+#include <string>
+#endif
+
+namespace KIM
+{
+namespace FILESYSTEM
+{
+class Path
+{
+ public:
+  Path() {}
+  Path(const char * str) : path_(str) {}
+  Path(const std::string & str) : path_(str) {}
+#ifdef KIM_API_USE_FILESYSTEM_LIBRARY
+  Path(const std::filesystem::path & p) : path_(p) {}
+  Path(std::filesystem::path && p) : path_(std::move(p)) {}
+#endif
+
+  Path & operator=(const std::string & other)
+  {
+    path_ = other;
+    return *this;
+  }
+
+  Path & operator=(const char * other)
+  {
+    path_ = other;
+    return *this;
+  }
+
+  Path & operator+=(char const * const s);
+  Path & operator/=(const Path & p);
+  Path operator/(const Path & p) const;
+  friend bool operator<(const Path & lhs, const Path & rhs)
+  {
+    return lhs.path_ < rhs.path_;
+  }
+  friend bool operator==(const Path & lhs, const Path & rhs)
+  {
+    return lhs.path_ == rhs.path_;
+  }
+
+  // Concatenates the current path and the argument.
+  Path & concat(const std::string & p);
```

Comment:
**General issue:** `std::` style convention vs. `KIM::` convention.  Since we're providing a class that is not going to be a simple drop-in replacement for C++17 `std::` features, I would generally prefer to conform to the `KIM::` naming conventions: `concat`->`Concatenate`, `string` -> `ToString`, `clear` -> `Clear`, `empty`->`Empty`, `filename`->`FileName`, `remove_filename`->`RemoveFileName`, `make_preferred`->`MakePreferred`, etc. Is there an argument for staying with the `std::` conventions?

Reply:
It's your choice, of course, but I advocate against renaming these methods. The idea was to replicate the interface of the `std::filesystem::path` class as closely as possible for the sake of forward compatibility. The aim still is to remove the `KIM::FILESYSTEM::Path` class at some point in the future, once the KIM API as a whole adopts C++17. This step will be easier if the interface is exactly replicated (as far as C++98 allows us). It also frees us from documenting these methods, because the ISO C++ standard defines their behaviour and names.